### ISSUE TITLE
ceph-exporter not cleaned up in v1.11.6

### ISSUE THREAD

**Issue by wjentner** [rook/rook#12249]
**Is this a bug report or feature request?**
* Bug Report

**Deviation from expected behavior:**
The deployment `rook-ceph-exporter-...`, service `rook-ceph-exporter`, and servicemonitor `rook-ceph-exporter` are still available.
Prometheus fires TargetDown for `serviceMonitor/rook-ceph/rook-ceph-exporter`. The alert `CephMgrIsAbsent` is firing and all of the `ceph_*` metrics are null.

**Expected behavior:**
Properly cleanup rook-ceph-exporter resources.

**How to reproduce it (minimal and precise):**

Updated from v1.11.5 to v1.11.6 without changing any variables (monitoring.enabled = true).

**File(s) to submit**:

* Cluster CR (custom resource), typically called `cluster.yaml`, if necessary

**Logs to submit**:

* Operator's logs, if necessary
* Crashing pod(s) logs, if necessary

  To get logs, use `kubectl -n <namespace> logs <pod name>`
  When pasting logs, always surround them with backticks or use the `insert code` button from the Github UI.
  Read [GitHub documentation if you need help](https://help.github.com/en/articles/creating-and-highlighting-code-blocks).

**Cluster Status to submit**:

* Output of krew commands, if necessary

  To get the health of the cluster, use `kubectl rook-ceph health`
  To get the status of the cluster, use `kubectl rook-ceph ceph status`
  For more details, see the [Rook Krew Plugin](https://rook.io/docs/rook/latest-release/Troubleshooting/krew-plugin)

**Environment**:
* OS (e.g. from /etc/os-release):
* Kernel (e.g. `uname -a`):
* Cloud provider or hardware configuration: bare-metal
* Rook version (use `rook version` inside of a Rook Pod): v1.11.6
* Storage backend version (e.g. for ceph do `ceph -v`): 17.2.6
* Kubernetes version (use `kubectl version`): 1.23.17
* Kubernetes cluster type (e.g. Tectonic, GKE, OpenShift): bare-metal / kubeadm
* Storage backend status (e.g. for Ceph use `ceph health` in the [Rook Ceph toolbox](https://rook.io/docs/rook/latest-release/Troubleshooting/ceph-toolbox/#interactive-toolbox)): HEALTH_OK

**Comment by reefland** [comment]
Same symptoms all ceph metrics are empty.  

Possibly related...  it looks like the service `rook-ceph-exporter` pod selectors, don't match the pods:

```
    selector:
      app: rook-ceph-exporter
      rook_cluster: rook-ceph
```

Try to use both labels:
```
$ k get pod -n rook-ceph -l app=rook-ceph-exporter,rook_cluster=rook-ceph
No resources found in rook-ceph namespace.
```

Try to use one label and it finds the pods:
```
$ k get pod -n rook-ceph -l app=rook-ceph-exporter
NAME                                        READY   STATUS    RESTARTS       AGE
rook-ceph-exporter-k3s01-85db995568-92skz   1/1     Running   3 (5d9h ago)   5d9h
rook-ceph-exporter-k3s02-5fd6b69c45-w8qzc   1/1     Running   0              21d
rook-ceph-exporter-k3s03-7cbf7b8db6-f2dw4   1/1     Running   4 (6d4h ago)   6d4h
rook-ceph-exporter-k3s04-86d7f9df87-q49g6   1/1     Running   3 (5d9h ago)   5d9h
rook-ceph-exporter-k3s06-5c4bfb546c-j8hnt   1/1     Running   4 (6d4h ago)   6d4h
```
(And one exporter for `k3s05` seems to be missing. )

The exporter pods do not have the `rook_cluster=rook-ceph` label:
```
$ k get pod rook-ceph-exporter-k3s02-5fd6b69c45-w8qzc -n rook-ceph -o jsonpath={.metadata.labels} | jq
{
  "app": "rook-ceph-exporter",
  "ceph-version": "17.2.6-0",
  "kubernetes.io/hostname": "k3s02",
  "node_name": "k3s02",
  "pod-template-hash": "5fd6b69c45",
  "rook-version": "v1.11.1"
}
```
Notice, wrong version is shown too.  This with label `rook_cluster=rook-ceph` seem to be getting updated:
```
$ kubectl -n rook-ceph get deployments -l rook_cluster=rook-ceph -o jsonpath='{range .items[*]}{.metadata.name}{"  \treq/upd/avl: "}{.spec.replicas}{"/"}{.status.updatedReplicas}{"/"}{.status.readyReplicas}{"  \trook-version="}{.metadata.labels.rook-version}{"\n"}{end}'
rook-ceph-crashcollector-k3s01          req/upd/avl: 1/1/1      rook-version=v1.11.6
rook-ceph-crashcollector-k3s02          req/upd/avl: 1/1/1      rook-version=v1.11.6
rook-ceph-crashcollector-k3s03          req/upd/avl: 1/1/1      rook-version=v1.11.6
rook-ceph-crashcollector-k3s04          req/upd/avl: 1/1/1      rook-version=v1.11.6
rook-ceph-crashcollector-k3s05          req/upd/avl: 1/1/1      rook-version=v1.11.6
rook-ceph-crashcollector-k3s06          req/upd/avl: 1/1/1      rook-version=v1.11.6
rook-ceph-mds-ceph-filesystem-a         req/upd/avl: 1/1/1      rook-version=v1.11.6
rook-ceph-mds-ceph-filesystem-b         req/upd/avl: 1/1/1      rook-version=v1.11.6
rook-ceph-mgr-a         req/upd/avl: 1/1/1      rook-version=v1.11.6
rook-ceph-mgr-b         req/upd/avl: 1/1/1      rook-version=v1.11.6
rook-ceph-mon-b         req/upd/avl: 1/1/1      rook-version=v1.11.6
rook-ceph-mon-e         req/upd/avl: 1/1/1      rook-version=v1.11.6
rook-ceph-mon-f         req/upd/avl: 1/1/1      rook-version=v1.11.6
rook-ceph-osd-0         req/upd/avl: 1/1/1      rook-version=v1.11.6
rook-ceph-osd-1         req/upd/avl: 1/1/1      rook-version=v1.11.6
rook-ceph-osd-2         req/upd/avl: 1/1/1      rook-version=v1.11.6
rook-ceph-osd-3         req/upd/avl: 1/1/1      rook-version=v1.11.6
rook-ceph-osd-4         req/upd/avl: 1/1/1      rook-version=v1.11.6
rook-ceph-osd-5         req/upd/avl: 1/1/1      rook-version=v1.11.6
rook-ceph-osd-6         req/upd/avl: 1/1/1      rook-version=v1.11.6
rook-ceph-osd-7         req/upd/avl: 1/1/1      rook-version=v1.11.6
rook-ceph-osd-8         req/upd/avl: 1/1/1      rook-version=v1.11.6
rook-ceph-rgw-ceph-objectstore-a        req/upd/avl: 1/1/1      rook-version=v1.11.6
```

Swap that label for `app=rook-ceph-exporter`:
```
$ kubectl -n rook-ceph get deployments -l app=rook-ceph-exporter -o jsonpath='{range .items[*]}{.metadata.name}{"  \treq/upd/avl: "}{.spec.replicas}{"/"}{.status.updatedReplicas}{"/"}{.status.readyReplicas}{"  \trook-version="}{.metadata.labels.rook-version}{"\n"}{end}'
rook-ceph-exporter-k3s01        req/upd/avl: 1/1/1      rook-version=v1.11.1
rook-ceph-exporter-k3s02        req/upd/avl: 1/1/1      rook-version=v1.11.1
rook-ceph-exporter-k3s03        req/upd/avl: 1/1/1      rook-version=v1.11.1
rook-ceph-exporter-k3s04        req/upd/avl: 1/1/1      rook-version=v1.11.1
rook-ceph-exporter-k3s06        req/upd/avl: 1/1/1      rook-version=v1.11.1
```

**Comment by travisn** [comment]
For those hitting this issue, does this workaround help?
- Delete the exporter deployments on each node
- Delete the exporter service
- Delete the exporter service monitor

@avanthakkar The reconciler for the exporter does not appear to remove the exporter and related resources if the ceph version changes. The cleanup of the exporter will only happen if the `hasCephPods` condition changes. The cleanup will also need to happen if the ceph version is not supported. 

Also, in addition to the exporter daemon, the related resources such as the service and service monitor would need to be deleted.

**Comment by reefland** [comment]
@travisn - I deploy via Helm / ArgoCD. Happy to report that manually deleting the items you stated do not get regenerated when the Helm output is reconciled so that is good. [for others -- main point of GitOps, if you delete something required/defined by the manifest files it will get automatically regenerated].

The  serviceMonitor `rook-ceph-mgr` , still not being picked up by Prometheus.

Not sure if the cause, but the service monitor has 3 label selectors:
```
$ k get servicemonitor rook-ceph-mgr -n rook-ceph -o jsonpath={.spec.selector} | jq
{
  "matchLabels": {
    "app": "rook-ceph-mgr",
    "mgr_role": "active",
    "rook_cluster": "rook-ceph"
  }
}
```

But the service only has 2 labels:
```
$ k get svc rook-ceph-mgr -n rook-ceph -o jsonpath={.metadata.labels} | jq

{
  "app": "rook-ceph-mgr",
  "rook_cluster": "rook-ceph"
}
```

Doesn't seem like the ServiceMonitor selector should be using selector `"mgr_role": "active"`, only the service should be using it to find the correct pod. And the service is using it:
```
$ k get svc rook-ceph-mgr -n rook-ceph -o jsonpath={.spec.selector} | jq
{
  "app": "rook-ceph-mgr",
  "mgr_role": "active",
  "rook_cluster": "rook-ceph"
}
```

**Comment by reefland** [comment]
Confirmed, I manually delete the selector `"mgr_role": "active"` from the serviceMonitor.

Within seconds Prometheus picked it up:
![image](https://github.com/rook/rook/assets/16230914/e746df9c-4616-49de-b1ca-9c8d756815f1)

**Comment by wjentner** [comment]
I tried the workaround and it works. However, after ~24 hours, the operator re-added the `mgr_role: active` label again to the selector. The helm deployment was not updated. It may be possible that the operator pod was restarted (maintenance on nodes / draining).

**Comment by reefland** [comment]
*Sigh*... yeah, as @wjentner states the labels comes back and breaks the serviceMonitor.

**Comment by travisn** [comment]
I see, the deletion of the unused exporter services was a distraction. I see now the regression because of #11845 that was included in v1.11.6 where the `mgr_role: active` label was added. I'll get a fix in for this shortly to remove the `mgr_role: active` label from the service monitor.

**Comment by travisn** [comment]
#12268 will fix the service monitor selector, but will leave this issue open for cleaning up the exporter resources.

**Comment by travisn** [comment]
To test the interim release build with the service monitor label fix, use the rook image: `rook/ceph:v1.11.6-11.g243401701`

@avanthakkar After you get a fix for cleaning up the exporter resources soon, then we can have a release.

**Comment by avanthakkar** [comment]
> To test the interim release build with the service monitor label fix, use the rook image: `rook/ceph:v1.11.6-11.g243401701`
> 
> @avanthakkar After you get a fix for cleaning up the exporter resources soon, then we can have a release.

I'm on it.

**Comment by reefland** [comment]
@travisn - tested the interim release image.  ServiceMonitor is fixed:
```
$ k get servicemonitor rook-ceph-mgr -n rook-ceph -o jsonpath={.spec.selector} | jq
{
  "matchLabels": {
    "app": "rook-ceph-mgr",
    "rook_cluster": "rook-ceph"
  }
}
```

Ceph dashboards are functional again.
